### DISCUSSION TITLE
Very slow response from the server when signup/login

### DISCUSSION THREAD

**Discussion by waldencui** [cookiecutter/cookiecutter-django#5230] [Q&A] [answered]
I'm using django-allauth to handle user signup/login. The current setup works, but the response from the server is incredibly slow (>15s, other types of responses are usually <500ms).

In base.py (I used cookiecutter-django to setup everything), the relevant settings are
```python
LOGIN_REDIRECT_URL = "myapp:home"
LOGIN_URL = "account_login"
ACCOUNT_SIGNUP_REDIRECT_URL = reverse_lazy("myapp:accept-terms-of-service")
ACCOUNT_EMAIL_REQUIRED = True
ACCOUNT_EMAIL_VERIFICATION = "mandatory"
ACCOUNT_ADAPTER = "project.users.adapters.AccountAdapter"
```
In adapters.py,
```python
class AccountAdapter(DefaultAccountAdapter):
    def is_open_for_signup(self, request: HttpRequest) -> bool:
        return getattr(settings, "ACCOUNT_ALLOW_REGISTRATION", True)

    def send_mail(self, template_prefix, email, context):
        mailing_thread = threading.Thread(
            target=super().send_mail,
            args=(template_prefix, email, context),
        )
        mailing_thread.start()
```
In views.py
```python
class CustomLoginView(LoginView):
    def form_valid(self, form):
        response = super().form_valid(form)
        user_id = self.request.user.id
        if not UserProfile.objects.filter(user_id=user_id).exists():
            return redirect("myapp:accept-terms-of-service")
        return response

def home(request):
    return render(request, "myapp/home.html")
```

The idea is to send a verification email to the user when he signs up. And when he first signs in, he will need to go to the process of accepting the legal documents.

The responses of two requests are very slow: 1. When the user is signing up, there is a significant delay before the server makes a response, though ultimately the user receives the correct email. 2. When the user first logs in, there is a significant delay before showing the accept terms page.

I have tried to use both django celery and threads to send the email, but the delay doesn't change at all (BTW all of them works, just very slow). So it's probably not a blocking problem. I am not sure if this is related to how boto3 is configured because it seems that AWS SES is ultimately sending the right email. But the following appears in the terminal each time when a user signs up:

```
DEBUG 2024-07-19 19:49:25,429 loaders 64407 140634015958720 Loading JSON file: /home/admin/.pyenv/versions/3.11.9/envs/django/lib/python3.11/site-packages/botocore/data/endpoints.json
DEBUG 2024-07-19 19:49:25,475 loaders 64407 140634015958720 Loading JSON file: /home/admin/.pyenv/versions/3.11.9/envs/django/lib/python3.11/site-packages/botocore/data/sdk-default-configuration.json
DEBUG 2024-07-19 19:49:25,496 loaders 64407 140634015958720 Loading JSON file: /home/admin/.pyenv/versions/3.11.9/envs/django/lib/python3.11/site-packages/botocore/data/sesv2/2019-09-27/service-2.json.gz
DEBUG 2024-07-19 19:49:25,522 loaders 64407 140634015958720 Loading JSON file: /home/admin/.pyenv/versions/3.11.9/envs/django/lib/python3.11/site-packages/botocore/data/sesv2/2019-09-27/endpoint-rule-set-1.json.gz
DEBUG 2024-07-19 19:49:25,523 loaders 64407 140634015958720 Loading JSON file: /home/admin/.pyenv/versions/3.11.9/envs/django/lib/python3.11/site-packages/botocore/data/partitions.json
DEBUG 2024-07-19 19:49:25,529 configprovider 64407 140634015958720 Looking for endpoint for sesv2 via: environment_service
DEBUG 2024-07-19 19:49:25,529 configprovider 64407 140634015958720 Looking for endpoint for sesv2 via: environment_global
DEBUG 2024-07-19 19:49:25,529 configprovider 64407 140634015958720 Looking for endpoint for sesv2 via: config_service
DEBUG 2024-07-19 19:49:25,529 configprovider 64407 140634015958720 Looking for endpoint for sesv2 via: config_global
DEBUG 2024-07-19 19:49:25,529 configprovider 64407 140634015958720 No configured endpoint found.
DEBUG 2024-07-19 19:49:25,532 loaders 64407 140634015958720 Loading JSON file: /home/admin/.pyenv/versions/3.11.9/envs/django/lib/python3.11/site-packages/botocore/data/_retry.json
DEBUG 2024-07-19 19:49:25,640 loaders 64407 140634052675264 Loading JSON file: /home/admin/.pyenv/versions/3.11.9/envs/django/lib/python3.11/site-packages/boto3/data/s3/2006-03-01/resources-1.json
DEBUG 2024-07-19 19:49:25,643 loaders 64407 140634052675264 Loading JSON file: /home/admin/.pyenv/versions/3.11.9/envs/django/lib/python3.11/site-packages/botocore/data/endpoints.json
DEBUG 2024-07-19 19:49:25,663 loaders 64407 140634052675264 Loading JSON file: /home/admin/.pyenv/versions/3.11.9/envs/django/lib/python3.11/site-packages/botocore/data/sdk-default-configuration.json
DEBUG 2024-07-19 19:49:25,680 loaders 64407 140634052675264 Loading JSON file: /home/admin/.pyenv/versions/3.11.9/envs/django/lib/python3.11/site-packages/botocore/data/s3/2006-03-01/service-2.json.gz
DEBUG 2024-07-19 19:49:25,701 loaders 64407 140634052675264 Loading JSON file: /home/admin/.pyenv/versions/3.11.9/envs/django/lib/python3.11/site-packages/botocore/data/s3/2006-03-01/endpoint-rule-set-1.json.gz
DEBUG 2024-07-19 19:49:25,709 loaders 64407 140634052675264 Loading JSON file: /home/admin/.pyenv/versions/3.11.9/envs/django/lib/python3.11/site-packages/botocore/data/partitions.json
DEBUG 2024-07-19 19:49:25,710 configprovider 64407 140634052675264 Looking for endpoint for s3 via: environment_service
DEBUG 2024-07-19 19:49:25,710 configprovider 64407 140634052675264 Looking for endpoint for s3 via: environment_global
DEBUG 2024-07-19 19:49:25,710 configprovider 64407 140634052675264 Looking for endpoint for s3 via: config_service
DEBUG 2024-07-19 19:49:25,710 configprovider 64407 140634052675264 Looking for endpoint for s3 via: config_global
DEBUG 2024-07-19 19:49:25,711 configprovider 64407 140634052675264 No configured endpoint found.
DEBUG 2024-07-19 19:49:25,714 loaders 64407 140634052675264 Loading JSON file: /home/admin/.pyenv/versions/3.11.9/envs/django/lib/python3.11/site-packages/botocore/data/_retry.json
DEBUG 2024-07-19 19:49:25,716 factory 64407 140634052675264 Loading s3:s3
DEBUG 2024-07-19 19:49:25,782 retryhandler 64407 140634015958720 No retry needed.
```

**Comment by medmunds** [comment]
[Note: cross-posted to pennersr/django-allauth#3978 and anymail/django-anymail#386.]

**Comment by foarsitter** [comment]
Are you using the debug toolbar? It can be slow with many queries. 

Attach a profiler and you will find the slowiness.

**Reply by waldencui** [reply]
I'm not using debug toolbar because this is in a production server. I'm not observing the same slowness during development.

**Reply by waldencui** [reply]
From the log it seems like certain queries are taking a lot of time:

> DEBUG 2024-07-21 08:01:55,510 http_protocol 138190 140482276182912 HTTP b'POST' request for ['127.0.0.1', 50410]
DEBUG 2024-07-21 08:01:55,877 utils 138190 140481993369280 (0.000) BEGIN; args=None; alias=default
DEBUG 2024-07-21 08:02:03,882 utils 138190 140481993369280 (0.006) SELECT "account_emailaddress"."id", "account_emailaddress"."user_id", "account_emailaddress"."email", "account_emailaddress"."verified", "account_emailaddress"."primary", "users_user"."password", "users_user"."last_login", "users_user"."is_superuser", "users_user"."is_staff", "users_user"."is_active", "users_user"."date_joined", "users_user"."id", "users_user"."is_hibernating", "users_user"."subscription_id", "users_user"."customer_id", "users_user"."name", "users_user"."email" FROM "account_emailaddress" INNER JOIN "users_user" ON ("account_emailaddress"."user_id" = "users_user"."id") WHERE "account_emailaddress"."email" = 'me@example.com'; args=('me@example.com',); alias=default
...
> DEBUG 2024-07-21 08:02:04,093 utils 138190 140481993369280 (0.001) SELECT 1 AS "a" FROM "account_emailaddress" WHERE ("account_emailaddress"."user_id" = '3774a6c1426b4173a2dcfc7826117796'::uuid AND "account_emailaddress"."verified") LIMIT 1; args=(Int4(1), UUID('3774a6c1-426b-4173-a2dc-fc7826117796')); alias=default
DEBUG 2024-07-21 08:02:11,966 utils 138190 140481993369280 (0.003) COMMIT; args=None; alias=default

But when I execute these queries directly in psql, they only take <10ms. How to explain such a large discrepancy?

**Reply by foarsitter** [reply]
The duration is before the query right? 
(0.000) BEGIN
(0.006) SELECT "account_emailaddress"."id"
(0.001) SELECT 1

**Reply by waldencui** [reply]
Oh, that's right. So in this case, the executions of these queries are actually very fast. It is something else that is causing the delay. How can I get more details on what's happening between these queries? This is the logging config that I use:
```python
LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "formatters": {
        "verbose": {
            "format": "%(levelname)s %(asctime)s %(module)s %(process)d %(thread)d %(message)s",
        },
    },
    "handlers": {
        "console": {
            "level": "DEBUG",
            "class": "logging.StreamHandler",
            "formatter": "verbose",
        },
        "file": {
            "level": "DEBUG",
            "class": "logging.FileHandler",
            "filename": "logs/django.log",
            "formatter": "verbose",
        },
    },
    "root": {"level": "DEBUG", "handlers": ["console", "file"]},
    "loggers": {
        "": {
            "handlers": ["file", "console"],
            "level": "DEBUG",
            "propagate": True,
        },
        "django": {
            "handlers": ["file", "console"],
            "level": "DEBUG",
            "propagate": False,
        },
        "django.request": {
            "handlers": ["file", "console"],
            "level": "DEBUG",
            "propagate": False,
        },
        "django.db.backends": {
            "handlers": ["file", "console"],
            "level": "DEBUG",
            "propagate": False,
        },
    },
}
```

**Reply by foarsitter** [reply]
If you think SES is causing it change the mail provider or use a smtp account. Another option is to use the production configuration on you dev machine. 

If you use sentry you can measure the performance of your app. Set the profiling and tracing params as described here: https://docs.sentry.io/platforms/python/integrations/django/#configure

**Reply by medmunds** [reply]
(One of the cross-posted threads concluded SES was not the issue.)

You mentioned (in a cross-post) you're using django-silk. That includes a SQL Inspection view where you can sort queries by execution time, which might help identify the problem. (But note that django-silk stores its measurements in your DB, so if you're running into DB performance issues django-silk might be making it worse—or could even be the root cause.)

There are several reasons your production performance might differ from dev. Are you on similar same hardware? Does your dev database have as much data as production?

Is your UserProfile.user_id a ForeignKey or OneToOneField (or otherwise indexed)? If not, that query in your CustomLoginView is going to get slower and slower the more users you add.

Or it might be unrelated to the DB. One of the cross-posted threads asked whether you're doing anything else during signup that would involve a network call—such as fetching profile info/photos from auth providers. Have you looked into this?

Have you checked your production error logs? When your mailing_thread raises an unhandled exception where does that error end up? (You're not handling it, and networks are never 100% reliable, so this is a "when" not an "if".) Any other errors messages around the time the delays occur?

**Reply by waldencui** [reply]
> If you think SES is causing it change the mail provider or use a smtp account. Another option is to use the production configuration on you dev machine.
> 
> If you use sentry you can measure the performance of your app. Set the profiling and tracing params as described here: https://docs.sentry.io/platforms/python/integrations/django/#configure

Thanks for the suggestion. I'll give sentry a shot.

**Comment by waldencui** [comment]
@medmunds @foarsitter The problem is due to RedisCache. I used ssh tunnels to test the local dev version of the same project on the server, and the slow page response issues never came up. After that, in my `production.py`, I replaced the following
```python
CACHES = {
    "default": {
        "BACKEND": "django_redis.cache.RedisCache",
        "LOCATION": env("REDIS_URL"),
        "OPTIONS": {
            "CLIENT_CLASS": "django_redis.client.DefaultClient",
            # Mimicing memcache behavior.
            # https://github.com/jazzband/django-redis#memcached-exceptions-behavior
            "IGNORE_EXCEPTIONS": True,
        },
    },
}
```
with the following
```python
CACHES = {
    "default": {
        "BACKEND": "django.core.cache.backends.locmem.LocMemCache",
        "LOCATION": "",
    },
}
```
and using the production setting, the problems went away. My understanding is that in production, it is preferable to use RedisCache rather than MemCache, but I still haven't figured out why in this case Redis is causing trouble.

**Reply by foarsitter** [reply, marked as answer]
I dont know how IGNORE_EXCEPTIONS exactly behaves but I think it hides the exception behind all this. 

Probably your redis url is invalid.

**Reply by waldencui** [reply]
Yes, it is indeed a Redis URL issue. Part of the reason is that I develop using local settings and in production, IGNORE_EXCEPTIONS is set to be True by cookiecutter-django. Therefore there're no debug messages. I'd suggest that IGNORE_EXCEPTIONS should be set to be False by default for developers to discover what's wrong easily.

**Reply by waldencui** [reply]
Big thanks to @medmunds and @foarsitter. I appreciate your help.
